### PR TITLE
Gangams/rs hyper scale 2022 ready

### PR DESCRIPTION
**Pull request by ganga1980**
This PR has following changes

- Fluentd Multiprocessor workers. 
- Agent will spin up the number of FLUEND workers depending on number of CPU cores 
- Implementation of Watch K8s clients to reduce number of K8s API calls and round-trip latencies
- In memory cache  kept up to date with the changes notified by watchers
- Kube Podinventory plugins into separate plugins to scale independently
- Separate plugin for Ingesting  POD MDM metrics 
- Separate plugin for KubePerf
- State being shared between the workers/processes via shared file 
- Make OUT MDM plugin work in multi-threaded environment

Validation done so far
    -  Validated AKS team's 5K cluster and ~50K pods
    -  Validated speech teams test and one prod cluster
    -  Validated on CI scale test cluster 

Attached CI scale cluster Perf run test reports with latest dev agent vs agent with 5k cluster improvements 

**Pending** 
  -   Waiting for at least one external large customer validation

### CODE REVIEW

**Comment by ganga1980** [comment]
**API Server Calls By Agent (latest dev agent vs  agent with 5k node cluster improvements)**
image
**Insights Ux with latest dev agent vs  agent with 5K node cluster improvements** 
latest dev agent 
image
agent with 5k node cluster improvements
image
**Agent Side latencies (latest dev agent vs 5k node cluster improvements agent)**
image
** RS CPU usage (latest dev agent vs  5k node cluster improvements agent) **
image
 **RS Memory Usage (latest dev agent vs  5k node cluster improvements agent)**
(latest dev agent vs  5k node cluster improvements agent)
** Ingestion data volume latest dev agent vs agent with 5K cluster improvements **
> Note - since no data drop in 5k cluster agent hence slight increase in data ingestion volume for KubePodInventory & Perf table 
image
This explains number of emitted records with latest dev agent vs agent with 5k cluster improvements
image

File: charts/azuremonitor-containers/templates/omsagent-deployment.yaml [hunk 1]
```diff
@@ -53,6 +53,11 @@ spec:
        resources:
 {{ toYaml .Values.omsagent.resources.deployment | indent 9 }}
        env:
+       - name: NUM_OF_FLUENTD_WORKERS
+         valueFrom:
+           resourceFieldRef:
+             containerName: omsagent
+             resource: limits.cpu
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply charts/azuremonitor-containers/templates/omsagent-deployment.yaml:60]: > this locks the num of fluentd workers to the num of cpu that is available. can their be a scenario where the Cx does not want this? Where they might want to assign only X number of CPUs?
> No, this is not the number of CPU cores available on the node rather number of CPU cores configured as limits on our Replica set  container. Currently this is 1 core and plan to use the VPA (once it available from AKS team) or pod resizer to scale our RS container limits based on the cluster size .

File: kubernetes/linux/main.sh [hunk 2]
```diff
@@ -90,6 +88,103 @@ checkAgentOnboardingStatus() {
       fi
 }
 
+setReplicaSetSpecificConfig() {
+      echo "num of fluentd workers:${NUM_OF_FLUENTD_WORKERS}"
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply kubernetes/linux/main.sh:92]: > can this variable be null or unset?

> No. This **NUM_OF_FLUENTD_WORKERS** is environment variable which defined in the yaml/chart. The value of this environment variable will be number CPU cores configured as limits on omsagent Replicat set container.

File: kubernetes/linux/main.sh [hunk 3]
```diff
@@ -90,6 +88,103 @@ checkAgentOnboardingStatus() {
...
+      export FLUENTD_FLUSH_INTERVAL="20s"
+      export FLUENTD_QUEUE_LIMIT_LENGTH="20" # default
+      export FLUENTD_KUBE_PERF_QUEUE_LIMIT_LENGTH="20"
+      export FLUENTD_MDM_FLUSH_THREAD_COUNT="5" # default
+      case $NUM_OF_FLUENTD_WORKERS in
+      5)
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply kubernetes/linux/main.sh]: > this case should capture >=5
> Yes, this already fixed. please look at the latest code.

File: kubernetes/linux/main.sh [hunk 4]
```diff
@@ -90,6 +88,103 @@ checkAgentOnboardingStatus() {
       fi
 }
 
+setReplicaSetSpecificConfig() {
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply kubernetes/linux/main.sh:91]: > overall comment: how were the values in the switch case decided? currently they appear to be adhoc. where they derived looking at results from performance testing?
> Fluentd Worker IDs , Flush interval, queue limits and max EPS determined based on how many cores are available as limit on the omsagent rs container. For example, higher the cores are available, the lesser (i.e. faster)  flush interval and similarly max EPS. Yes, these being tested as part of the performance testing.

File: source/plugins/ruby/out_mdm.rb [hunk 5]
```diff
@@ -165,6 +158,10 @@ def start
       end
     end
 
+    def multi_workers_ready?
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply source/plugins/ruby/out_mdm.rb:161]: > is this function needed if this is always true?

> This is contract of Fluentd to tell that, this particular plugin is multi worker ready meaning  this plugin can run on multiple of CPU cores (aka workers) if multi cores are available. Here are some references - 

https://www.fluentd.org/blog/fluentd-v0.14.12-has-been-released

**Plugin MUST have #multi_workers_ready? method and return true**
The default return value of this method is true in filter/buffer/parser/formatter/storage plugins
But it's false in input/output plugins
**3rd party plugins SHOULD re-implemnet #multi_workers_ready? to return true to support multi process workers**

https://www.rubydoc.info/gems/fluentd/0.14.15/Fluent%2FPlugin%2FFileBuffer:multi_workers_ready%3F

File: source/plugins/ruby/in_kube_podmdminventory.rb [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        $log.debug_backtrace(errorStr.backtrace)
+        ApplicationInsightsUtility.sendExceptionTelemetry(errorStr)
+      end
+    end
+
+    def parse_and_emit_records()
+      begin
+        $log.info "in_kube_podmdminventory:parse_and_emit_records:Start:getMDMRecords @ #{Time.now.utc.iso8601}"
+        mdmPodRecords = getMDMRecords()
+        mdmPodRecordItems =
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply source/plugins/ruby/in_kube_podmdminventory.rb]: > this looks incorrect?
> this was already fixed, looks like you have looked at before the fix. please double check on latest code.

File: source/plugins/ruby/podinventory_to_mdm.rb [hunk 7]
```diff
@@ -129,7 +129,7 @@ def get_pod_inventory_mdm_records(batch_time)
             controllerNameDimValue: podControllerNameDimValue,
             podCountMetricValue: value,
           }
-          records.push(JSON.parse(record))
+          records.push(Yajl::Parser.parse(record))
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply source/plugins/ruby/podinventory_to_mdm.rb:132]: > Should there be a StringIO here?
> Here we are parsing very small metric record payload. StringIO helps if parsing larger payloads in streaming fashion.

File: source/plugins/ruby/KubernetesApiClient.rb [hunk 8]
```diff
@@ -456,19 +508,16 @@ def getContainerResourceRequestsAndLimits(pod, metricCategory, metricNameToColle
...
-              metricItems.push(metricProps)             
+              metricItems.push(metricProps)
               #No container level limit for the given metric, so default to node level limit
             else
-              nodeMetricsHashKey = clusterId + "/" + nodeName + "_" + "allocatable" + "_" + metricNameToCollect
-              if (metricCategory == "limits" && @@NodeMetrics.has_key?(nodeMetricsHashKey))
-                metricValue = @@NodeMetrics[nodeMetricsHashKey]
-                #@Log.info("Limits not set for container #{clusterId + "/" + podUid + "/" + containerName} using node level limits: #{nodeMetricsHashKey}=#{metricValue} ")
-                               
+              if (metricCategory == "limits" && !nodeAllocatableRecord.nil? && !nodeAllocatableRecord.empty? && nodeAllocatableRecord.has_key?(metricNameToCollect))
```
[the remark this answers is not in the document]
Comment (ganga1980) [reply source/plugins/ruby/KubernetesApiClient.rb:519]: > can this if clause we simplied?
>  this only place and this logic being used hence moving to helper function will not help. btw, I dont see any complexity or lengthy.

[the remark this answers is not in the document]
Comment (pfrcks) [reply source/plugins/ruby/KubernetesApiClient.rb:519]: complexity is only in understanding not in time complexity. this was a nit so resolving

**Review by pfrcks — APPROVED** [review]

**Review by pfrcks — APPROVED** [review]

**Pull request merged**